docs/development/test_code.md

```diff
@@ -0,0 +1,62 @@
+# Chapter 4: Understanding testing style and structure
+
+Tests for the extension and the language server are stored separately in their
+respective directories inside the `test` sub-folders.
+
+Throughout the project and for all types of tests, we have:
+
+1. **Test fixtures:** Settings and Ansible files against which the tests are
+   run.
+2. **Test scripts:** the actual test code.
+
+## Extension tests
+
+There are three types of tests for the extension:
+
+1. **Unit tests:** Check various functions used across the extension.
+2. **End-to-end (e2e) tests:** Test the entire user flow. They open a test host
+   of the extension, run the whole workflow from launching the extension,
+   loading Ansible files in the editor, activating the language server, using
+   the language features, and finally asserting the outcome.
+3. **UI tests:** Check the presence and functionality of various UI elements in
+   the VS Code interface, from the status bar and activity bar to web-views and
+   terminal views.
+
+## Language server tests
+
+The language server has only unit tests in its directory, as the end-to-end
+testing is already covered by the extension's e2e tests. A test file for each
+provider (language feature) is present, that tests the functionality of the
+language feature.
+
+## Running the Tests
+
+Each type of test has its own script and can be run by `yarn run <script-name>`:
+
+!!! tip
+
+    Always run `yarn run compile` before running the tests.
+
+### Extension test scripts
+
+1. **unit-tests:** Runs the unit tests for the extension.
+2. **test-e2e:** Runs the end-to-end tests for the extension.
+3. **test-ui-current:** Runs the UI tests on the latest version of VS Code.
+4. **test-ui-oldest:** Runs the UI tests on the oldest supported version of VS
+   Code.
+
+!!! note
+
+    For the following tests (UI tests for the extension), you need to package and build a `.vsix file` of the extension by running `yarn run package`. As you run the UI tests, the .vsix file is automatically installed and the UI elements are checked.
+
```

Comment:
Could add a note about MOCHA_GREP for UI tests, optional.

Reply:
Could you give me more info about it that I can add?

Reply:
You can call specific UI tests with the envvar MOCHA_GREP, like so:
`MOCHA_GREP="Test One Click Trial feature" yarn coverage-ui-with-mock-lightspeed-server` or 
`MOCHA_GREP="Verify the execution of playbook using ansible-navigator command" yarn coverage-ui-current` so you don't have to run all the tests in that set. I believe this is only for UI tests that use Mocha, but Tami could confirm more. I also tagged you in the slack thread where he explained it to me for more details.

Reply:
Changes LGTM!